ygen/protoelements.go

```diff
@@ -116,14 +117,94 @@ func (s *genState) yangTypeToProtoScalarType(args resolveTypeArgs) (*mappedType,
+		}, nil
+	case yang.Yunion:
+		return s.protoUnionType(args, basePackageName, enumPackageName)
 	default:
 		// TODO(robjs): implement missing types.
-		//	- identityref
 		//	- binary
 		//	- bits
-		//	- union
-		return nil, fmt.Errorf("unimplemented type: %s", args.yangType.Kind)
+		return nil, fmt.Errorf("unimplemented type in scalar generation: %s", args.yangType.Kind)
+	}
+}
+
+// protoUnionType resolves the types that are included within the YangType in resolveTypeArgs into the
+// scalar type that can be included in a protobuf oneof. The basePackageName and enumPackageName are used
+// to determine the paths that are used for enumerated types within the YANG schema. Each union is
+// resolved into a oneof that contains the scalar types, for example:
+//
+// leaf a {
+//	type union {
+//		type string;
+//		type int32;
+//	}
+// }
+//
+// Is represented in the output protobuf as:
+//
+// oneof a {
+//	string a_string = NN;
+//	int32 a_int32 = NN;
+// }
+//
+// The mappedType's unionTypes can be output through a template into the oneof.
+func (s *genState) protoUnionType(args resolveTypeArgs, basePackageName, enumPackageName string) (*mappedType, error) {
+	var errs []error
+	unionTypes := make(map[string]int)
+	// Extract the subtypes that are defined into map, keyed on the union type.
+	for _, subtype := range args.yangType.Type {
+		errs = append(errs, s.protoUnionSubTypes(subtype, args.contextEntry, unionTypes, basePackageName, enumPackageName)...)
+	}
+
+	if errs != nil {
+		return nil, fmt.Errorf("errors mapping element: %v", errs)
+	}
+
+	return &mappedType{unionTypes: unionTypes}, nil
+}
+
+// protoUnionSubTypes extracts all possible subtypes of a YANG union. It returns a map keyed by the mapped type
+// along with any errors that occur. The context entry is used to map types when the leaf that the type is associated
+// with is required for mapping. The currentType map is updated as an in-out argument. The basePackageName and enumPackageName
+// are used to map enumerated typedefs and identityrefs to the correct type. It returns a slice of errors if they occur
+// mapping subtypes.
+func (s *genState) protoUnionSubTypes(subtype *yang.YangType, ctx *yang.Entry, currentTypes map[string]int, basePackageName, enumPackageName string) []error {
+	var errs []error
+	// If subtype.Type is not empty, then this is a union itself.
+	if isUnionType(subtype) {
+		for _, st := range subtype.Type {
+			errs = append(errs, s.protoUnionSubTypes(st, ctx, currentTypes, basePackageName, enumPackageName)...)
+		}
+		return errs
 	}
+
+	var mtype *mappedType
+	switch subtype.Kind {
+	case yang.Yidentityref:
+		// Handle the case that the context entry is not the correct entry to deal with. This occurs when the subtype is
+		// an identityref.
+		mtype = &mappedType{nativeType: fmt.Sprintf("%s.%s.%s", basePackageName, enumPackageName, s.identityrefBaseTypeFromIdentity(subtype.IdentityBase))}
```

Comment:
If possible it would be simpler to always call identityrefBaseTypeFromIdentity (which could be called something like resolveIdentifyRef) that returns the resolved type if it's an identityref, or the passed in type if it's not. Then you can always call yangTypeToProtoScalarType with the resolved type.

Reply:
The issue with this is that this moves the complexity inside that type resolution (which is already relatively complex as it has naming logic in it). 

What is it that you're trying to fix with this suggestion please?

Reply:
if you follow the suggestion, you end up with this pattern:
newYANGType := resolveIdentityRef()
mtype = yangTypeToProtoScalarType()
rather than 
switch {
 case union:
   // do X + Y
 default:
   // do Y
}
Just seems simpler to me overall (even with the added check inside resolveIdentityRef). 
But it's just a suggestion, feel free to take it or leave it.

Reply:
The `identityrefBaseTypeFromIdentity` just returns a string, it doesn't return a `yang.YangType`. In the issue #33, I'll discuss how we can clean this up with you.